modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentChildInnerHitContextBuilder.java

```diff
@@ -0,0 +1,259 @@
+            JoinFieldInnerHitSubContext joinFieldInnerHits = new JoinFieldInnerHitSubContext(name, context, typeName, joinFieldMapper);
+            setupInnerHitsContext(queryShardContext, joinFieldInnerHits);
+            innerHitsContext.addInnerHitDefinition(joinFieldInnerHits);
+        } else {
+            if (innerHitBuilder.isIgnoreUnmapped() == false) {
+                throw new IllegalStateException("no join field has been configured");
+            }
+        }
+    }
+
+    private void handleParentFieldInnerHits(SearchContext context, InnerHitsContext innerHitsContext) throws IOException {
+        QueryShardContext queryShardContext = context.getQueryShardContext();
+        DocumentMapper documentMapper = queryShardContext.documentMapper(typeName);
+        if (documentMapper == null) {
+            if (innerHitBuilder.isIgnoreUnmapped() == false) {
+                throw new IllegalStateException("[" + query.getName() + "] no mapping found for type [" + typeName + "]");
+            } else {
+                return;
+            }
+        }
+        String name = innerHitBuilder.getName() != null ? innerHitBuilder.getName() : documentMapper.type();
+        ParentChildInnerHitSubContext parentChildInnerHits = new ParentChildInnerHitSubContext(
+            name, context, queryShardContext.getMapperService(), documentMapper
+        );
+        setupInnerHitsContext(queryShardContext, parentChildInnerHits);
+        innerHitsContext.addInnerHitDefinition(parentChildInnerHits);
+    }
+
+    static final class JoinFieldInnerHitSubContext extends InnerHitsContext.InnerHitSubContext {
+        private final String typeName;
+        private final ParentJoinFieldMapper joinFieldMapper;
+
+        JoinFieldInnerHitSubContext(String name, SearchContext context, String typeName, ParentJoinFieldMapper joinFieldMapper) {
+            super(name, context);
+            this.typeName = typeName;
+            this.joinFieldMapper = joinFieldMapper;
+        }
+
+        @Override
+        public TopDocs[] topDocs(SearchHit[] hits) throws IOException {
+            Weight innerHitQueryWeight = createInnerHitQueryWeight();
+            TopDocs[] result = new TopDocs[hits.length];
+            for (int i = 0; i < hits.length; i++) {
+                SearchHit hit = hits[i];
+                SearchHitField joinField = hit.getFields().get(joinFieldMapper.name());
+                if (joinField == null) {
+                    result[i] = Lucene.EMPTY_TOP_DOCS;
+                    continue;
+                }
+
+                String parentIdFieldName = ParentJoinFieldMapper.getParentIdFieldName(joinFieldMapper.name(), typeName);
+                SearchHitField parentIdField = hit.getFields().get(parentIdFieldName);
+                QueryShardContext qsc = context.getQueryShardContext();
+
+                Query q;
+                if (parentIdField == null) {
```

Comment:
Does this work if the document is a child **and** a parent ? I think it's simpler if we use the context directly.
`hasParent` expects a `child` and builds `parent` inner hits. `hasChildren` expects a `parent` and builds `children` inner hits ?